google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -52,26 +80,52 @@ opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeParentSpan(
                             opentelemetry::common::AttributeValue>>;
   auto constexpr kMaxOtelLinks = 128;
   std::vector<std::pair<SpanContext, AttributesList>> links;
-  auto batch_size = message_spans.size();
+  auto batch_size = static_cast<std::int64_t>(message_spans.size());
   links.reserve(batch_size);
   // If the batch size is less than the max size, add the links to a single
-  // span.
-  if (batch_size < kMaxOtelLinks) {
-    std::transform(
-        message_spans.begin(), message_spans.end(), std::back_inserter(links),
-        [i = static_cast<std::int64_t>(0)](auto const& span) mutable {
-          return std::make_pair(
-              span->GetContext(),
-              AttributesList{{"messaging.pubsub.message.link", i++}});
-        });
+  // span. If the batch size is greater than the max size, this will be a parent
+  // span with no links and each child spans will contain links.
+  bool const is_small_batch = batch_size <= kMaxOtelLinks;
+  if (is_small_batch) {
+    GenerateLinks(message_spans.begin(), message_spans.end(), links);
   }
+
   auto batch_sink_parent_span =
       internal::MakeSpan("BatchSink::AsyncPublish",
                          /*attributes=*/
                          {{"messaging.pubsub.num_messages_in_batch",
                            static_cast<std::int64_t>(batch_size)}},
                          /*links*/ links);
 
+  std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
```

Comment:
Feels like this could be simplified.  Maybe starting with:

https://godbolt.org/z/Yjcv7cdoT

Reply:
Tried to integrate the comment. It could be cleaner though...

Reply:
It is much better, thanks.  Yes, it feels like we should be able to simplify further.  Let me think about this tomorrow morning.

Reply:
Maybe something along these lines. What do you think?

https://godbolt.org/z/MM74G5Kn5

Reply:
Tried to get it a little neater.

I used type aliases since we're in the anonymous namespace and i think that helps a little. Lmk if there's anything that reads weird still.